20251028-BRS-workflow-versioning.md

```diff
@@ -0,0 +1,416 @@
+        }
+        
+        await context.CreateTimer(TimeSpan.FromMinutes(15));
+        context.ContinueAsNew(id, false);
+        
+        return null;
+    }
+}
+```
+
+Because of the deterministic nature of Dapr Workflows, we have several considerations that must be taken into account
+when considering a home-made versioning system here:
+- **One cannot simply swap out this workflow type in a new deployment:** Swapping out the workflow type might fail
+catastrophically if inlight-workflows were unable to complete or if a replay were performed because of a previously-
+encountered failure. This would lead to developer confusion and they might assume that a new deployment would supersede
+previous deployment versions.
+- **Code cannot simply be added to the workflow**: Following the reasoning of the previous point, adding new code 
+violates deterministic integrity as it changes the logic that provides and uses the inputs and outputs at each of the
+invocation boundaries (again, there's a singular exception to this noted in the Caveat section below).
+
+If the workflow were to have stopped when it was finished, it would give the developer an opportunity to relaunch it 
+using a different type enabling a blue/green-like deployment model - that is, deploy a new "green" type until all the 
+existing "blue" types are known to have stopped running and are instead running as "green" types", then delete the "blue"
+types from the next deployment altogether. However, in this very likely scenario, the developer never has an opportunity
+to swap out Workflow types, especially in those long-running types that have been idle for some time even between new
+blue/green deployments.
+
+Now, had the developer thought about this in a previous version, they had options to implement this themselves. They 
+could have:
+- Introduced an activity that performed some reflective analysis of its own service, or
+- Used some sort of Workflow registry to determine that there was a newer type available to swap to.
+
+It could them have scheduled a one-time Dapr JOb with the updated Workflow type name as the payload and ended the current
+workflow to have the job invoke the new Workflow type. However, this introduces a lot of complexity and still doesn't
+address how bugs in live workflows can be fixed between migrations to new types. Since we're trying to simplify distributed
+application development with Dapr, I propose that there's a richer and simpler approach here. I introduce a
+powerful mechanism to swap between Dapr Workflow types and introduce patches to outstanding workflows with minimal 
+effort by existing Dapr developers.
+
+## Implementation Details
+I propose that the internal Dapr Workflow API be slightly modified to include a string value whenever a workflow is
+executed. This value contains the name of the versioned type that ran for any given workflow invocation as well as a 
+list of each named patch applied and run in that execution.
```

Comment:
I think it'd be useful to incorporate the implementation details from #92 as I think they could pretty well for both type selection and patch enablement.

It would need the following changes:
- Add workflow type to the `OrchestratorResponse` and `OrchestratorStartedEvent`. This way the SDK will tell the runtime which specific type it used, and it'll saved in the workflow event history.
- Adjust it to contain a list of 'enabled' patches instead of the branch versions.

This way, when the SDK receives the events from the runtime it'll know which specific type needs to run (if that version has been set in the first `OrchestratorStartedEvent`), and also if a patch is enabled (if the corresponding `OrchestratorStartedEvent` has it enabled).

Reply:
I agree - Josh raises a good point below about concerns regarding putting seemingly arbitrary constraints on the patch names. While I have them to ensure that version parsing works without incident, by instead using a dictionary, we can remove those concerns and clean up the type a little better.

While I added the protos today, I'll revisit to accommodate your structure instead.